### PR TITLE
Add functions to get all of fixed|free, scaled parameter values

### PR DESCRIPTION
**Pull request by yannikschaelte**

### CODE REVIEW

File: petab/problem.py [hunk 1]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
         return measurements.get_noise_distributions(
             measurement_df=self.measurement_df)
 
+    def _to_mask(v: List, free: bool = True, fixed: bool = True):
+        """Apply mask of only free or only fixed values."""
+        if not free and not fixed:
+            raise ValueError("Disabling both `free` and `fixed` would "
```
Comment (dweindl) [inline petab/problem.py]: Just return an empty list? :)

File: petab/problem.py [hunk 2]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
         return measurements.get_noise_distributions(
             measurement_df=self.measurement_df)
 
+    def _to_mask(v: List, free: bool = True, fixed: bool = True):
```
Comment (dweindl) [inline petab/problem.py]: What's `v`?

File: petab/problem.py [hunk 3]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        for free parameters."""
+        return self.get_x_nominal(fixed=False, scaled=True)
+
+    @property
+    def x_nominal_fixed_scaled(self) -> List:
+        """Parameter table nominal values with applied parameter scaling,
+        for fixed parameters."""
+        return self.get_x_nominal(free=False, scaled=True)
+
+    def get_lb(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Docstring?

File: petab/problem.py (hunk 2 above)
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Method should have "self" as first argument](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 4]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
 
+    def _to_mask(v: List, free: bool = True, fixed: bool = True):
+        """Apply mask of only free or only fixed values."""
+        if not free and not fixed:
+            raise ValueError("Disabling both `free` and `fixed` would "
+                             "yield an empty list.")
+        if not free:
+            return [v[ix] for ix in self.x_fixed_indices]
+        if not fixed:
+            return [v[ix] for ix in self.x_free_indices]
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Value 'v' is unsubscriptable](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Undefined variable 'self'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 5]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
-        return list(parameters.map_scale(self.parameter_df[LOWER_BOUND],
-                                         self.parameter_df[PARAMETER_SCALE]))
+        return self.get_lb(scaled=True)
+
+    def get_ub(self, free: bool = True, fixed: bool = True,
+               scaled: bool = False):
+        v = self.parameter_df[UPPER_BOUND]
+        if scaled:
+            v = parameters.map_scale(v, self.parameter_df[PARAMETER_SCALE])
+        return self._to_mask(v, free=free, fixed=fixed)
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Argument 'free' passed by position and keyword in method call](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 6]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
         return measurements.get_noise_distributions(
             measurement_df=self.measurement_df)
 
+    def _to_mask(v: List, free: bool = True, fixed: bool = True):
+        """Apply mask of only free or only fixed values."""
+        if not free and not fixed:
+            raise ValueError("Disabling both `free` and `fixed` would "
+                             "yield an empty list.")
+        if not free:
+            return [v[ix] for ix in self.x_fixed_indices]
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Value 'v' is unsubscriptable](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 7]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        if not fixed:
+            return [v[ix] for ix in self.x_free_indices]
+        return v
+
+    def get_x_ids(self, free: bool = True, fixed: bool = True):
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 8]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        return self.get_lb()
 
     @property
     def lb_scaled(self) -> List:
         """Parameter table lower bounds with applied parameter scaling"""
-        return list(parameters.map_scale(self.parameter_df[LOWER_BOUND],
-                                         self.parameter_df[PARAMETER_SCALE]))
+        return self.get_lb(scaled=True)
+
+    def get_ub(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 9]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
-        return list(estimated.index.values)
+        """Parameter table parameter IDs, for free parameters."""
+        return self.get_x_ids(fixed=False)
+
+    @property
+    def x_fixed_ids(self) -> List[str]:
+        """Parameter table parameter IDs, for fixed parameters."""
+        return self.get_x_ids(free=False)
+
+    def get_x_nominal(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py (hunk 6 above)
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Undefined variable 'self'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py (hunk 3 above)
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

**Comment by yannikschaelte** [comment]
> Do we really need both the properties and the functions? Can live with it, but seems a bit bloated.

yeah, don't need the properties, but did not want to destroy backward compatibility in amici+pypesto atm

File: petab/problem.py (hunk 2 above)
Comment (yannikschaelte) [reply petab/problem.py]: any vector. maybe `arr` instead

File: petab/problem.py [hunk 10]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        if not free and not fixed:
+            raise ValueError("Disabling both `free` and `fixed` would "
+                             "yield an empty list.")
+        if not free:
+            return [v[ix] for ix in self.x_fixed_indices]
+        if not fixed:
+            return [v[ix] for ix in self.x_free_indices]
+        return v
+
+    def get_x_ids(self, free: bool = True, fixed: bool = True):
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 11]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        return self.get_lb()
 
     @property
     def lb_scaled(self) -> List:
         """Parameter table lower bounds with applied parameter scaling"""
-        return list(parameters.map_scale(self.parameter_df[LOWER_BOUND],
-                                         self.parameter_df[PARAMETER_SCALE]))
+        return self.get_lb(scaled=True)
+
+    def get_ub(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 12]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
+        for free parameters."""
+        return self.get_x_nominal(fixed=False, scaled=True)
+
+    @property
+    def x_nominal_fixed_scaled(self) -> List:
+        """Parameter table nominal values with applied parameter scaling,
+        for fixed parameters."""
+        return self.get_x_nominal(free=False, scaled=True)
+
+    def get_lb(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py [hunk 13]
```diff
@@ -347,66 +347,111 @@ def get_noise_distributions(self):
...
-        return list(estimated.index.values)
+        """Parameter table parameter IDs, for free parameters."""
+        return self.get_x_ids(fixed=False)
+
+    @property
+    def x_fixed_ids(self) -> List[str]:
+        """Parameter table parameter IDs, for fixed parameters."""
+        return self.get_x_ids(free=False)
+
+    def get_x_nominal(self, free: bool = True, fixed: bool = True,
```
Comment (dweindl) [inline petab/problem.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4968439)

File: petab/problem.py (hunk 2 above)
Comment (dweindl) [reply petab/problem.py]: Not so much about the name, but document arguments.

**Pull request merged**